File: 206.Reverse-Linked-List.go
// https://leetcode.com/problems/reverse-linked-list/description/
//
// algorithms
// Easy (50.50%)
// Total Accepted:    448.8K
// Total Submissions: 888.7K
// beats 100.0% of golang submissions

package leetcode

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseList(head *ListNode) *ListNode {
	tmp := &ListNode{}

	for {
		if head == nil {
			break
		}
		nextNode := head.Next
		head.Next = tmp.Next
		tmp.Next = head
		head = nextNode
	}

	return tmp.Next
}

File: 206.Reverse-Linked-List.java
// https://leetcode.com/problems/reverse-linked-list/
//
// algorithms
// Easy (53.87%)
// Total Accepted:    557,002
// Total Submissions: 1,034,002
// beats 100.0% of java submissions


/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        
        ListNode res = new ListNode(0);
        ListNode tmp = null;
        
        while (head != null) {
            tmp = head.next;
            head.next = res.next;
            res.next = head;
            head = tmp;
        }
        
        return res.next;
    }
}